runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2625,3 +2625,186 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
+
+      arrayCheckBlock->append(TR::TreeTop::create(comp, arrayCheckNode, NULL, NULL));
+      cfg->addEdge(callBlock, newCallBlock);
+      }
+
+   // Insert adjust trees
+   TR::Node* adjustedNode = TR::TransformUtil::generateDataAddrLoadTrees(comp, node->duplicateTree());
+   TR::Node *newStore = TR::Node::createStore(symRef, adjustedNode);
+   TR::TreeTop *newStoreTree = TR::TreeTop::create(comp, newStore);
+   adjustBlock->append(newStoreTree);
+
+   return newCallBlock;
+   }
+
+void J9::TransformUtil::transformUnsafeCopyMemorytoArrayCopyForOffHeap(TR::Compilation *comp, TR::TreeTop *arrayCopyTT, TR::Node *arraycopyNode)
+   {
+   // When using balanced GC policy with offheap allocation enabled, there are three possible for an argument type:
+   //
+   // 1.) The type is known to be a non-array object at compile time. In this scenario, the final address
+   //     can be calculated by simply adding ref and offset.
+   // 2.) The type is known to be an array at compile time. In this scenario, if the ref at runtime is `null` then
+   //     final address is calculated as in case A. If not `null` then final address is the adjusted ref, by loading
+   //     the dataAddr pointer field then add it to the offset.
+   // 3.) The type of the object at dest is unknown at compile time (type is `java/lang/Object`). In this scenario,
+   //     a runtime null check and arrayCHK must be generated to determine whether it needs to be handled such as
+   //     case A or case B.
+
+   TR::Node *src        = arraycopyNode->getChild(1);
+   TR::Node *dest       = arraycopyNode->getChild(3);
+
+   // Check src/dest type at compile time
+   int srcSigLen, destSigLen;
+   const char *srcObjTypeSig = src->getSymbolReference() ? src->getSymbolReference()->getTypeSignature(srcSigLen) : 0;
+   const char *destObjTypeSig = dest->getSymbolReference() ? dest->getSymbolReference()->getTypeSignature(destSigLen) : 0;
+
+   // Case 3
+   bool srcArrayCheckNeeded, destArrayCheckNeeded;
+   if (!srcObjTypeSig)
+      destArrayCheckNeeded = true;
+   else
+      {
+      srcArrayCheckNeeded = false;
+      TR_OpaqueClassBlock *srcClass = comp->fej9()->getClassFromSignature(srcObjTypeSig, srcSigLen, comp->getCurrentMethod());
+      srcArrayCheckNeeded |= srcClass ? TR::Compiler->cls.isInterfaceClass(comp, srcClass) : true;
+      srcArrayCheckNeeded |= strncmp(srcObjTypeSig, "Ljava/lang/Object;", 18 /*strlen*/) == 0;
+      }
+
+   if (!destObjTypeSig)
+      destArrayCheckNeeded = true;
+   else
+      {
+      destArrayCheckNeeded = false;
+      TR_OpaqueClassBlock *destClass = comp->fej9()->getClassFromSignature(destObjTypeSig, destSigLen, comp->getCurrentMethod());
+      destArrayCheckNeeded |= destClass ? TR::Compiler->cls.isInterfaceClass(comp, destClass) : true;
+      destArrayCheckNeeded |= strncmp(destObjTypeSig, "Ljava/lang/Object;", 18 /*strlen*/) == 0;
+      }
```

Comment:
This logic is repeated and now seems involved enough that it might be worth factoring out into a helper function. It's up to you, but I'd recommend it considering the copies seem to want to change in tandem

Reply:
After your suggestion to use
```
destArrayCheckNeeded = destClass == NULL
   || destClass == comp->getObjectClassPointer()
   || TR::Compiler->cls.isInterfaceClass(comp, destClass);
```
Refactoring would use a helper with 4 arguments to in place of (technically) 2 line of codes. Which doesn't seem worthwhile hence I'm skipping this.